internal/credential/vault/repository_credential_library.go

```diff
@@ -207,24 +320,221 @@ func (r *Repository) UpdateCredentialLibrary(ctx context.Context, scopeId string
+			dmp.setUsernameAttribute(nmp.UsernameAttribute)
+		}
+		if mappingOverrideMask[passwordAttributeField] {
+			dmp.setPasswordAttribute(nmp.PasswordAttribute)
+		}
+		if dmp.UsernameAttribute == "" && dmp.PasswordAttribute == "" {
+			return noChange, nil
+		}
+		return insertFn(dmp), nil
+	case om != nil && nm == nil:
+		// delete attributes in dbmask
+		dm := om.clone()
+		dmp, ok := dm.(*UserPasswordOverride)
+		if !ok {
+			return nil, fmt.Errorf("unexpected mapping override type %T: %v", dm, dm)
+		}
+		var mask []string
+		if mappingOverrideMask[usernameAttributeField] {
+			mask = append(mask, usernameAttributeField)
+			dmp.deleteUsernameAttribute()
+		}
+		if mappingOverrideMask[passwordAttributeField] {
+			mask = append(mask, passwordAttributeField)
+			dmp.deletePasswordAttribute()
+		}
+		if dmp.UsernameAttribute == noOverride && dmp.PasswordAttribute == noOverride {
+			return deleteFn(dmp), nil
+		}
+		return updateFn(dmp, mask), nil
+	case om != nil && nm != nil:
+		// delete or update
+		dm := om.clone()
+		dmp, ok := dm.(*UserPasswordOverride)
+		if !ok {
+			return nil, fmt.Errorf("unexpected mapping override type %T: %v", dm, dm)
+		}
+		nmp, ok := nm.(*UserPasswordOverride)
+		if !ok {
+			return nil, fmt.Errorf("unexpected mapping override type %T: %v", nm, nm)
+		}
+		var mask []string
+		if mappingOverrideMask[usernameAttributeField] {
+			mask = append(mask, usernameAttributeField)
+			switch nmp.UsernameAttribute {
+			case "":
+				dmp.deleteUsernameAttribute()
+			default:
+				dmp.setUsernameAttribute(nmp.UsernameAttribute)
+			}
+		}
+		if mappingOverrideMask[passwordAttributeField] {
+			mask = append(mask, passwordAttributeField)
+			switch nmp.PasswordAttribute {
+			case "":
+				dmp.deletePasswordAttribute()
+			default:
+				dmp.setPasswordAttribute(nmp.PasswordAttribute)
+			}
+		}
+		if dmp.UsernameAttribute == noOverride && dmp.PasswordAttribute == noOverride {
+			return deleteFn(dmp), nil
+		}
+		return updateFn(dmp, mask), nil
+	}
+	return noChange, nil
+}
+
+func noChange(_ context.Context, _ db.Writer) (*oplog.Message, error) {
+	return nil, nil
+}
+
 // LookupCredentialLibrary returns the CredentialLibrary for publicId.
 // Returns nil, nil if no CredentialLibrary is found for publicId.
 func (r *Repository) LookupCredentialLibrary(ctx context.Context, publicId string, _ ...Option) (*CredentialLibrary, error) {
 	const op = "vault.(Repository).LookupCredentialLibrary"
 	if publicId == "" {
 		return nil, errors.New(ctx, errors.InvalidParameter, op, "no public id")
 	}
-	l := allocCredentialLibrary()
+	l := allocPublicLibrary()
 	l.PublicId = publicId
 	if err := r.reader.LookupByPublicId(ctx, l); err != nil {
 		if errors.IsNotFoundError(err) {
 			return nil, nil
 		}
 		return nil, errors.Wrap(ctx, err, op, errors.WithMsg(fmt.Sprintf("failed for: %s", publicId)))
 	}
-	return l, nil
+	return l.toCredentialLibrary(), nil
+}
+
+type publicLibrary struct {
```

Comment:
can you add docs for this? There is a nice comment on corresponding sql view, but some docs here would be nice as well.